Leaderboard command: Migrate to slash commands

Mclilzee opened the pull request:
<!-- Thank you for taking the time to contribute to The Odin Project. In order to get this pull request (PR) merged in a reasonable amount of time, you must complete this entire template. -->

## Because
<!-- Summarize the purpose or reasons for this PR, e.g. what problem it solves or what benefit it provides. -->
Update Odin bot old command `!leaderboard` into the new slash command `/leaderboad` which has 2 sub commands, one for fetching server ranking, and the second for fetching a specific user's points. The latter will fix an old broken command `!points` which was supposed to show a specific user's current points.

## This PR
<!-- A bullet point list of one or more items describing the specific changes. -->
- Replace `!leaderboard` command with `/leaderboard` slash command.
- Integrate testing to the new slash commands
- Provide sub command `/leaderboard user` which take a user as required option, then provide points for that specific user
- Provide sub command `/leaderboard ranking` which have 2 Optional parameters, limit and offset, returns the server ranking for all users.
- Remove old broken command `!points` create tests for it and add it as the subcommand `/leaderboard user`
- Cleanup snapshots and tests, to work with the new commands
- Added subdirectory __tests__ which will be holding the tests for `new-era-commands` and add it as ignored in `index.js`'s glob

## Issue
<!--
If this PR closes an open issue in this repo, replace the XXXXX below with the issue number, e.g. Closes #2013.

If this PR closes an open issue in another TOP repo, replace the #XXXXX with the URL of the issue, e.g. Closes https://github.com/TheOdinProject/curriculum/issues/XXXXX

If this PR does not close, but is related to another issue or PR, you can link it as above without the 'Closes' keyword, e.g. 'Related to #2013'.
-->
Closes #356

## Additional Information
<!-- Any other information about this PR, such as a link to a Discord discussion. -->
The new changes will work with the current API. No changes to the API required, but some changes could be beneficial.

The changes right now will fetch a user's format properly and show the users' points, but it will not show the user's Rank. In order for that to work, I need to make a change on the API level as mentioned previously in #356 . When the API change happens, this update will automatically reflect the changes, no further changes required to be done in the leaderboard command. Unless the goal is to refactor and clean up.

## Pull Request Requirements
<!-- Replace the whitespace between the square brackets with an 'x', e.g. [x]. After you create the PR, they will become checkboxes that you can click on. -->
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Callbacks command: Update verbiage`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If this PR adds new features or functionality, I have added new tests
-   [x] If applicable, I have ensured all tests related to any command files included in this PR pass, and/or all snapshots are up to date

Mclilzee commented:
@ManonLef I would appreciate your input on this, you have always been good with designs and I feel my naming and descriptions could be improved.

ManonLef commented:
@Mclilzee I'd love to but I haven't worked with the points yet. Feel free to add me to a test server or link me the points-bot / token guide (if any) so I can jump into it later!

Mclilzee commented:
@01zulfi I hope we are not giving you too much work for this bot. If you can please give this a quick review. @ManonLef had a brilliant idea to implement. Instead of having the `leaderboard` and `points` commands separate we can combine them into :
- `/points help`
- `/points leaderboard'
- `/points user`
- 
So I will be working on migrating that.

If you have any input as well on the testing methods that I have used, I wasn't sure how to go about testing the slash commands, so I implemented the best way I came up with.

Also further more about the 3 points i said earlier in the issue, after further consideration the change of API I provided will not work. It is true it show the correct rank for a specific member, but we don't have yet a way to filter `active` from `inactive` members, so the rank is most likely gonna be off from what the `leaderboard` actually shows. Right now the bot will just show points without a rank, in order to show a rank my only option for now is to make the call for all data and filter them inside the bot itself finding my user + the index as a rank.

Which further more makes the `/api/points/{discord_id}` dead code in the backend, any input on that?

ManonLef commented:
>  @ManonLef had a brilliant idea to implement. Instead of having the leaderboard and points commands separate we can combine them 

Needs this context I added : Not saying it's a good option but we could explore it. See what's possible when it comes to grouping them.

😂

Mclilzee commented:
It's great option!

01zulfi reviewed (changes requested):
@Mclilzee 🔥 🔥 🔥 

On an initial glance, this looks great. Also, kudos for adding tests, this is an area where we are lacking.

I have a suggestion regarding code structure. What do you think about moving all of the functions used into a separate `LeaderboardService`, or `PointsService` class in the `services/leaderboard` directory? And then in the slash command file, just call a method(s) from that class. We can also move the tests there.

Mclilzee commented:
@01zulfi I can do that certainly, I can move the functions into `services/points/points.service.js` and `points.service.test.js` That will certainly simplify the testing process into actual unit tests.

I still need your input about how to proceed with getting the rank, so far the only way right now is to loop through all records of the leaderboard. Should I proceed?

01zulfi commented:
> I still need your input about how to proceed with getting the rank, so far the only way right now is to loop through all records of the leaderboard. Should I proceed?

Yup!

Mclilzee commented:
@01zulfi This should be ready to give it a test. Special thanks to @ManonLef couldn't have done it without my twin ❤️ 

I tried to keep the class consistent with the other services, I had `handleInteraction` static method inside the service. But I didn't feel this was the right way to do it, I would have personally separated the Service from handling the interaction and their commands. Services to only return embed data instead of calling the reply themselves. Which in return will change the way I could handle the unit tests without mocking the Interaction heavily.

I didn't want to change the way it is now for consistency, but that can be a new issue opened afterward to separate Services from Interaction handling, for all related Services.

01zulfi commented on botCommands/__snapshots__/question-legacy.test.js.snap:
Not sure why this file is added in this PR?

01zulfi commented on botCommands/__snapshots__/simpleCommands.test.js.snap:
```diff
@@ -43,29 +43,6 @@ exports[`!lenny callback returns correct output 1`] = `"( ͡° ͜ʖ ͡°)"`;
 
 exports[`!motivate callback returns correct output 1`] = `"Don't give up! https://www.youtube.com/watch?v=KxGRhd_iWuE"`;
 
-exports[`!question callback returns correct output 1`] = `
-{
-  "embeds": [
-    {
-      "color": 13407555,
-      "description": "
-Asking context-rich questions makes it easy to receive help, and makes it easy for others to help you quickly! Great engineers ask great questions, and the prompt below is an invitation to improve your skills and set yourself up for success in the workplace. 
-
-**Project/Exercise:**
-**Lesson link:**
-**Code:** [code sandbox like replit or codepen]
-**Issue/Problem:** [screenshots if applicable]
-**What I expected:** 
-**What I've tried:** 
-
-For even more context around on how to hone your question-asking skills, give this a read: https://www.theodinproject.com/guides/community/how_to_ask
-              ",
-      "title": "Asking Great Questions",
```
Not sure about this change either, I might lack context for these changes.

01zulfi commented on botCommands/question-legacy.js:
Again, why has this file been added?

01zulfi commented on botCommands/question-legacy.test.js:
Same here

01zulfi commented on botCommands/simpleCommands.js:
```diff
@@ -28,30 +28,6 @@ const fu = {
 };
 registerBotCommand(fu.regex, fu.cb);
 
-const question = {
-  regex: /(?<!\S)!question(?!\S)/,
-  cb: () => {
-    const questionEmbed = new Discord.EmbedBuilder()
-      .setColor('#cc9543')
-      .setTitle('Asking Great Questions')
-      .setDescription(`
-Asking context-rich questions makes it easy to receive help, and makes it easy for others to help you quickly! Great engineers ask great questions, and the prompt below is an invitation to improve your skills and set yourself up for success in the workplace. 
-
-**Project/Exercise:**
-**Lesson link:**
-**Code:** [code sandbox like replit or codepen]
```
?

01zulfi commented on botCommands/simpleCommands.test.js:
```diff
@@ -247,68 +247,6 @@ describe('!fu', () => {
   });
 });
 
-describe('!question', () => {
-  describe('regex', () => {
```
?

01zulfi commented on new-era-commands/index.js:
```diff
@@ -1,14 +1,15 @@
 const path = require('path');
 
-const commandFiles = globSync('./new-era-commands/**/*.js', { ignore: 'new-era-commands/*.js' });
+const commandFiles = globSync('./new-era-commands/**/*.js', { ignore: ['new-era-commands/*.js', 'new-era-commands/__tests__/**'] });
```
Since we moved all our tests from this directory, I think it's safe to revert this change.

01zulfi commented on new-era-commands/slash/freelance.js:
Not sure here either

01zulfi commented on services/points/points.service.js:
```diff
@@ -0,0 +1,125 @@
+    const user = interaction.options.getUser('name');
+    const response = await axios.get(PointsService.API_URL);
+
+    // eslint-disable-next-line
+    const data = response.data.filter((member) => interaction.guild.members.cache.get(member.discord_id));
+
```
I'm not sure why we're filtering for users that exist in the server. Won't this make the ranks in the leaderboard flaky? Since if a person with points leaves the server, it would change the leaderboard rank for everyone below them. I might have missed something and this could be the intended behavior. Please let me know your thoughts, thanks!

01zulfi commented on services/points/points.service.js:
```diff
@@ -0,0 +1,125 @@
+  static async displayLeaderboard(interaction) {
+    const response = await axios.get(PointsService.API_URL);
+    // eslint-disable-next-line max-len
+    const users = response.data.filter((user) => interaction.guild.members.cache.get(user.discord_id));
```
Left comment on filtering above.

01zulfi commented on services/points/points.service.js:
```diff
@@ -0,0 +1,125 @@
+
+    let offset = interaction.options.getInteger('offset');
+    offset = offset > 0 ? offset : 0;
+    // Always show the last members if offset too high
+    offset = offset + limit < users.length ? offset : Math.max(0, users.length - limit);
```
Nicely handled 💪

01zulfi commented on services/points/points.service.test.js:
Pretty elaborate test suite!

01zulfi reviewed (changes requested):
Sorry for the laaaate review.

Everything is lookin' pretty good. There's some unrelated changes I'm not sure about, please do provide context for them. Also left a couple comments!

Mclilzee commented:
@01zulfi There might have been confusion with the rebase I did earlier from main which included some files, I will explore them later today and clean it up with rebase -i. 

As for the filtering that was the default behavior we currently have, if a user leaves the server their rank will be ignored and the person below them will take on their rank and so on. Was this not the behavior expected? I feel it would be unfair for someone's rank to be lower than a person who is not currently on the server, or has been banned.

01zulfi commented:
@Mclilzee Thanks for the update. Let's keep the filtering then!

Mclilzee commented:
@01zulfi I updated my rebase and removed the commit you requested. I'm sorry for giving you trouble with my old rebase mess up, I should have checked it myself.

I'm quite sure my problem was rebasing from upstream main rather than pulling the newest changes into my origin main and then rebasing from there. All the commits that have already been merged with upstream main are now discarded from this PR.

If there is anything else you need changing, please let me know.
PS: If you know what messed up my rebase the first time, let me know too. This would be an educational moment for me <3

01zulfi commented:
> PS: If you know what messed up my rebase the first time, let me know too. This would be an educational moment for me <3

Honestly I have no idea, git can be confusing at times lol 😄

01zulfi reviewed (approved):
Let's get this merged 🚀
